### PR TITLE
Big refactoring of the Core

### PR DESCRIPTION
**Pull request by psfinaki**
Apparently this is too big for any proper perception so I will just describe the results achieved:
1. The `WikiParsing` project is extracted from the `Core` project. It works directly with Wikipedia,  parses the contents, provides string outputs and is the only thing dependent on `FSharp.Data`.
2. The `Common` project is also extracted from `Core`, it contains super basic things like extension methods and grammar categories used everywhere else in the code.
3. Tests are split as per the new code structure.
4.  `Core` itself is a smaller mess now:
core
Part of speech folders (`Noun`, `Adjective`, `Verb`) include all the functionality within each part of speech. The last modules (`Noun`, `Adjective`, `Verb`) are kind of containers to collect all the info  from those various underlying modules.
5. The exercise models (`NounPlural`, `VerbImperative`, ...) now live in their own folder and contain only pure Azure Table Storage entities. Validation is moved to a separate places (modules in the `Validation` folder) and is called from the very top thing (the `Scraper` project).

Apparently there is more to be done - encapsulation, cleaner Azure Storage serialization and various  minor cleanups are on the way. But I decided to stop here for now - tests are passing, I rechecked everything manually to verify I did not break anything.

### CODE REVIEW

File: src/Core/Adjectives/Adjective.fs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿module Adjective
```
Comment (reflash) [inline src/Core/Adjectives/Adjective.fs]: Is this file related to parsing the word? Seems it does call adjective parser

File: src/Core/Nouns/FeminineNounPatternDetector.fs.fs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿module FeminineNounPatternDetector
 
-open Declensions
+open NounArticle
```
Comment (reflash) [inline src/Core/Nouns/FeminineNounPatternDetector.fs.fs]: Are the detectors called from the client code or are they used during the wiki scraping process?

File: src/Core/Validation/AdjectiveValidation.fs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+let hasRequiredInfoComparative =
+    tryGetAdjective
+    >> Option.exists (hasChildPart "stupňování")
+
+let hasRequiredInfoPlural = 
+    tryGetAdjective 
+    >> Option.exists (hasChildPart "skloňování")
+
+let isValidCommon = isPositive
```
Comment (reflash) [inline src/Core/Validation/AdjectiveValidation.fs]: Not sure what this name means, a bit unclear for me

File: src/Core/Verbs/VerbClasses.fs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿module VerbClasses
+
+open StringHelper
+open VerbArticle
+open Reflexives
+
+let getClassByThirdPersonSingular = function
+    | form when form |> ends "á"  -> 5
```
Comment (reflash) [inline src/Core/Verbs/VerbClasses.fs]: Do we have names for these classes? Maybe we should use discriminated union instead of plain integers?

File: src/Server/Server.fsproj [hunk 5]
```diff
@@ -1,9 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project Sdk="Microsoft.NET.Sdk.Web">
-    <PropertyGroup>
-        <OutputType>Exe</OutputType>
-        <TargetFramework>netcoreapp2.1</TargetFramework>
-    </PropertyGroup>
+  <PropertyGroup>
+      <OutputType>Exe</OutputType>
+      <TargetFramework>netcoreapp3.0</TargetFramework>
```
Comment (reflash) [inline src/Server/Server.fsproj]: So that means we're on net core 3 now. Have you tried running it on non-Windows machine? Last time I tried core3 it was a bit unstable with F#

File: src/Core/Adjectives/Adjective.fs (hunk 1 above)
Comment (psfinaki) [reply src/Core/Adjectives/Adjective.fs]: Well, it is just a container of all the stuff we need to know about adjectives, you can think of it as an Adjective class, just did not want to bring real classes here.

File: src/Core/Nouns/FeminineNounPatternDetector.fs.fs (hunk 2 above)
Comment (psfinaki) [reply src/Core/Nouns/FeminineNounPatternDetector.fs.fs]: Only during scraping. Client does not any processing and only grabs stuff from the database.

File: src/Core/Validation/AdjectiveValidation.fs (hunk 3 above)
Comment (psfinaki) [reply src/Core/Validation/AdjectiveValidation.fs]: So, some adjectives are valid for the `Plural` exercise, other are valid for the `Comparative` exercise, most are actually valid for both. Thus we have different validation functions and common validation steps are included in the `isValidCommon` function. In this case, there is only one common check but there can be more stuff (consider `isValidCommon` in `NounValidation`).

File: src/Core/Nouns/FeminineNounPatternDetector.fs.fs (hunk 2 above)
Comment (reflash) [reply src/Core/Nouns/FeminineNounPatternDetector.fs.fs]: So this detector is not used from Server components? Also notice .fs.fs in the file name

File: src/Core/Validation/AdjectiveValidation.fs (hunk 3 above)
Comment (reflash) [reply src/Core/Validation/AdjectiveValidation.fs]: Can it be named `isValid` or `isValidAdjective`? You can keep `isValidCommon` if those don't seem to fit

File: src/Core/Verbs/VerbClasses.fs (hunk 4 above)
Comment (psfinaki) [reply src/Core/Verbs/VerbClasses.fs]: Well, official grammar addresses them just by [numbers](https://cs.wikipedia.org/wiki/Česká_slovesa#Slovesné_třídy_a_vzory). Although it might still be good idea, e.g. for uniforming serialization and overall approach. However, I would take it separately.

File: src/Server/Server.fsproj (hunk 5 above)
Comment (psfinaki) [reply src/Server/Server.fsproj]: Well, a month ago it they made it [generally available](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0/) (before it was a preview), so I would expect it being stable. I did not try it myself though.

File: src/Core/Nouns/FeminineNounPatternDetector.fs.fs (hunk 2 above)
Comment (psfinaki) [reply src/Core/Nouns/FeminineNounPatternDetector.fs.fs]: No, it's not used from there. Thanks, name fixed

File: src/Core/Verbs/VerbClasses.fs (hunk 4 above)
Comment (reflash) [reply src/Core/Verbs/VerbClasses.fs]: Okay, could you create an issue for that?

File: src/Core/Validation/AdjectiveValidation.fs (hunk 3 above)
Comment (psfinaki) [reply src/Core/Validation/AdjectiveValidation.fs]: Good idea, did it in the neighboring modules as well

File: src/Core/Adjectives/Adjective.fs (hunk 1 above)
Comment (psfinaki) [reply src/Core/Adjectives/Adjective.fs]: Yes. Server uses only the `Storage` module and the exercise models (table entities).
But regretfully those entities are not POCOs yet, they depend on the `Core` stuff to set the props. That setting logic should be abstracted away and so eventually `Server` will not depend on the fat core but only on a small subset of it.

But for now I chunked away at least the validation part from "models".
